servant-client/test/Servant/MiddlewareSpec.hs

```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-name-shadowing #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -freduction-depth=100 #-}
+
+module Servant.MiddlewareSpec (spec) where
+
+import Control.Arrow
+  ( left,
+  )
+import Control.Concurrent (newEmptyMVar, putMVar, takeMVar)
+import Control.Monad.IO.Class
+import Data.ByteString.Builder (toLazyByteString)
+import Data.Monoid ()
+import Prelude.Compat
+import Servant.Client
+import Servant.Client.Core (RequestF (..))
+import Servant.Client.Internal.HttpClient (ClientMiddleware)
+import Servant.ClientTestUtils
+import Test.Hspec
+import Prelude ()
+import Control.Exception (Exception, throwIO, try)
+import Data.IORef (newIORef, modifyIORef, readIORef)
+
+spec :: Spec
+spec = describe "Servant.MiddlewareSpec" $ do
+  middlewareSpec
+
+runClientWithMiddleware :: ClientM a -> ClientMiddleware -> BaseUrl -> IO (Either ClientError a)
+runClientWithMiddleware x mid baseUrl' =
+  runClientM x ((mkClientEnv manager' baseUrl') {middleware = mid})
+
+data CustomException = CustomException deriving (Show, Eq)
+instance Exception CustomException
+
+middlewareSpec :: Spec
+middlewareSpec = beforeAll (startWaiApp server) $ afterAll endWaiApp $ do
+  describe "middleware" $ do
+    it "Raw request and response can be accessed in middleware" $ \(_, baseUrl) -> do
```

Comment:
suggestion: remove  `describe` calls, there's only one per `it` statement and either it does not add any information or it could be incorporated in the `it "..."` description